// @egret/eui-compiler/lib/emitter/javascript-emitter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var codegen = require("escodegen");
var _1 = require(".");
var JavaScriptEmitter = /** @class */ (function (_super) {
    __extends(JavaScriptEmitter, _super);
    function JavaScriptEmitter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.mapping = {
            number: createNumberOrBooleanLiteral,
            boolean: createNumberOrBooleanLiteral,
            string: createStringLiteral,
            any: createStringLiteral,
            'egret.Rectangle': createNewRectangle,
            'object': _this.createNewObject.bind(_this)
        };
        _this.javascript = '';
        _this.body = [];
        return _this;
    }
    JavaScriptEmitter.prototype.getResult = function () {
        return this.javascript;
    };
    JavaScriptEmitter.prototype.emitHeader = function (themeData) {
        var outputText = '';
        outputText += extendsHelper;
        outputText += euiHelper(themeData.skins);
        this.javascript += outputText;
    };
    JavaScriptEmitter.prototype.emitSkinNode = function (filename, skinNode) {
        var ast = this.generateJavaScriptAST(skinNode);
        var text = codegen.generate(ast);
        this.javascript += text + "\n";
        this.javascript += "\ngenerateEUI.paths['" + filename + "'] = " + skinNode.namespace + "." + skinNode.classname + ";\n        ";
    };
    JavaScriptEmitter.prototype.generateJavaScriptAST = function (skinNode) {
        var _this = this;
        var ids = [];
        var states = [];
        if (skinNode.states) {
            for (var _i = 0, _a = skinNode.states; _i < _a.length; _i++) {
                var stateName = _a[_i];
                states.push({ name: stateName, items: [] });
            }
        }
        function visitChildren(node) {
            if (node.id) {
                ids.push(node.id);
            }
            var _loop_1 = function (stateAttribute) {
                var arr = states.find(function (s) { return s.name === stateAttribute.name; });
                arr.items.push(Object.assign({}, stateAttribute, { context: node.varIndex }));
            };
            for (var _i = 0, _a = node.stateAttributes; _i < _a.length; _i++) {
                var stateAttribute = _a[_i];
                _loop_1(stateAttribute);
            }
            node.children.forEach(visitChildren);
        }
        visitChildren(skinNode);
        var className = createIdentifier(skinNode.classname);
        var namespace = createIdentifier(skinNode.namespace);
        this.writeToBody(emitSkinPart(ids));
        var context = createIdentifier('_this');
        this.emitAttributes(context, skinNode);
        this.emitChildren(context, skinNode);
        if (skinNode.states.length > 0) {
            this.writeToBody(createExpressionStatment(createAssignmentExpression(createMemberExpression(context, createIdentifier('states')), createArray(states.map(function (s) {
                return createNewExpression(createMemberExpression(createIdentifier('eui'), createIdentifier("State")), [
                    createStringLiteral(s.name),
                    createArray(s.items.map(function (item) {
                        if (item.type === 'set') {
                            return createNewExpression(createMemberExpression(createIdentifier("eui"), createIdentifier("SetProperty")), [
                                createStringLiteral("a" + item.context),
                                createStringLiteral(item.attribute.key),
                                _this.mapping[item.attribute.type](item.attribute.value)
                            ]);
                        }
                        else {
                            return createNewExpression(createMemberExpression(createIdentifier("eui"), createIdentifier("AddItems")), [
                                createStringLiteral("a" + item.context),
                                createStringLiteral(""),
                                createNumberOrBooleanLiteral(1),
                                createStringLiteral("")
                                // createStringLiteral(item.attribute.key),
                                // this.mapping[item.attribute.type](item.attribute.value)
                            ]);
                        }
                    }))
                ]);
            })))));
        }
        var code = createExpressionStatment(createAssignmentExpression(createMemberExpression(namespace, className), createClass(className, this.body)));
        return createProgram([code]);
    };
    JavaScriptEmitter.prototype.emitNode = function (node) {
        var context = createVarIndexIdentifier(node);
        this.writeToBody(emitCreateNode(context, emitComponentName(node.type)));
        if (node.id) {
            this.writeToBody(createExpressionStatment(createAssignmentExpression(createMemberExpression(createIdentifier("_this"), createIdentifier(node.id)), context)));
        }
        if (node.stateAttributes.length > 0) {
            this.writeToBody(createExpressionStatment(createAssignmentExpression(createMemberExpression(createIdentifier("_this"), context), context)));
        }
        this.emitAttributes(context, node);
        this.emitChildren(context, node);
    };
    JavaScriptEmitter.prototype.emitChildren = function (context, node) {
        if (node.children.length == 0) {
            return;
        }
        for (var _i = 0, _a = node.children; _i < _a.length; _i++) {
            var child = _a[_i];
            this.emitNode(child);
        }
        this.writeToBody(emitElementsContent(context.name, node.children.map(createVarIndexIdentifier)));
    };
    JavaScriptEmitter.prototype.emitAttributes = function (context, node) {
        for (var _i = 0, _a = node.attributes; _i < _a.length; _i++) {
            var attribute = _a[_i];
            this.writeToBody(this.emitAttribute(context, attribute));
        }
        ;
    };
    JavaScriptEmitter.prototype.createNewObject = function (value) {
        var varIndexIdentifer = createIdentifier("a" + value.varIndex);
        this.emitNode(value);
        return varIndexIdentifer;
    };
    JavaScriptEmitter.prototype.writeToBody = function (node) {
        this.body.push(node);
    };
    JavaScriptEmitter.prototype.emitAttribute = function (context, attribute) {
        var emitterFunction = this.mapping[attribute.type];
        if (!emitterFunction) {
            console.error("找不到", attribute.key, attribute.type, attribute.value);
            process.exit(1);
            return null;
        }
        else {
            return createExpressionStatment(createAssignmentExpression(createMemberExpression(context, createIdentifier(attribute.key)), emitterFunction(attribute.value)));
        }
    };
    return JavaScriptEmitter;
}(_1.BaseEmitter));
exports.JavaScriptEmitter = JavaScriptEmitter;
function createVarIndexIdentifier(node) {
    return createIdentifier("a" + node.varIndex);
}
function emitComponentName(type) {
    var arr = type.split('.');
    return createMemberExpression(createIdentifier(arr[0]), createIdentifier(arr[1]));
}
function emitElementsContent(context, ids) {
    return createExpressionStatment(createAssignmentExpression(createMemberExpression(createIdentifier(context), createIdentifier("elementsContent")), createArray(ids)));
}
function emitCreateNode(varIndex, componentName) {
    return {
        type: "VariableDeclaration",
        declarations: [
            {
                type: "VariableDeclarator",
                id: varIndex,
                init: createNewExpression(componentName, [])
            }
        ],
        kind: "var"
    };
}
function emitSkinPart(skins) {
    return createExpressionStatment(createAssignmentExpression(createMemberExpression(createIdentifier("_this"), createIdentifier("skinParts")), createArray(skins.map(createStringLiteral))));
}
function createArray(elements) {
    return {
        type: "ArrayExpression",
        elements: elements
    };
}
function createStringLiteral(value) {
    return {
        type: "Literal",
        value: value,
        raw: "\"" + value + "\""
    };
}
function createNumberOrBooleanLiteral(value) {
    if (typeof value === 'number') {
        if (value < 0 || (value === 0 && 1 / value < 0)) {
            return {
                "type": "UnaryExpression",
                "operator": "-",
                "argument": {
                    type: "Literal",
                    value: -value,
                    raw: -value.toString()
                },
                "prefix": true
            };
        }
    }
    return {
        type: "Literal",
        value: value,
        raw: value.toString()
    };
}
function createNewRectangle(value) {
    var args = value.split(",").map(function (v) { return parseInt(v); });
    return createNewExpression(createMemberExpression(createIdentifier('egret'), createIdentifier('Rectangle')), [
        createNumberOrBooleanLiteral(args[0]),
        createNumberOrBooleanLiteral(args[1]),
        createNumberOrBooleanLiteral(args[2]),
        createNumberOrBooleanLiteral(args[3])
    ]);
}
function createNewExpression(callee, args) {
    return {
        type: "NewExpression",
        callee: callee,
        arguments: args
    };
}
function createExpressionStatment(expression) {
    return {
        "type": "ExpressionStatement",
        "expression": expression
    };
}
function createAssignmentExpression(left, right) {
    return {
        "type": "AssignmentExpression",
        "operator": "=",
        "left": left,
        "right": right
    };
}
function createMemberExpression(object, property) {
    return {
        "type": "MemberExpression",
        "computed": false,
        "object": object,
        "property": property
    };
}
function createIdentifier(name) {
    return {
        "type": "Identifier",
        name: name
    };
}
function createProgram(body) {
    return {
        "type": "Program",
        body: body,
        "sourceType": "script"
    };
}
function createClass(className, constractorBody) {
    var superCall = {
        "type": "VariableDeclaration",
        "declarations": [
            {
                "type": "VariableDeclarator",
                "id": {
                    "type": "Identifier",
                    "name": "_this"
                },
                "init": {
                    "type": "LogicalExpression",
                    "operator": "||",
                    "left": {
                        "type": "CallExpression",
                        "callee": {
                            "type": "MemberExpression",
                            "computed": false,
                            "object": {
                                "type": "Identifier",
                                "name": "_super"
                            },
                            "property": {
                                "type": "Identifier",
                                "name": "call"
                            }
                        },
                        "arguments": [
                            {
                                "type": "ThisExpression"
                            }
                        ]
                    },
                    "right": {
                        "type": "ThisExpression"
                    }
                }
            }
        ],
        "kind": "var"
    };
    var returnStatement = {
        "type": "ReturnStatement",
        "argument": createIdentifier("_this")
    };
    var fullConstractorBody = [superCall].concat(constractorBody).concat([returnStatement]);
    var functionArguments = [
        createMemberExpression(createIdentifier("eui"), createIdentifier("Skin"))
    ];
    return {
        "type": "CallExpression",
        "callee": {
            "type": "FunctionExpression",
            "id": null,
            "params": [
                {
                    "type": "Identifier",
                    "name": "_super"
                }
            ],
            "body": {
                "type": "BlockStatement",
                "body": [
                    {
                        "type": "ExpressionStatement",
                        "expression": {
                            "type": "CallExpression",
                            "callee": {
                                "type": "Identifier",
                                "name": "__extends"
                            },
                            "arguments": [
                                className,
                                {
                                    "type": "Identifier",
                                    "name": "_super"
                                }
                            ]
                        }
                    },
                    {
                        "type": "FunctionDeclaration",
                        "id": className,
                        "params": [],
                        "body": {
                            "type": "BlockStatement",
                            "body": fullConstractorBody
                        },
                        "generator": false,
                        "expression": false,
                        "async": false
                    },
                    {
                        "type": "ReturnStatement",
                        "argument": className
                    }
                ]
            },
            "generator": false,
            "expression": false,
            "async": false
        },
        "arguments": functionArguments
    };
}
var extendsHelper = "\nvar __extends = (this && this.__extends) || (function () {\n    var extendStatics = function (d, b) {\n        extendStatics = Object.setPrototypeOf ||\n            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||\n            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };\n        return extendStatics(d, b);\n    };\n    return function (d, b) {\n        extendStatics(d, b);\n        function __() { this.constructor = d; }\n        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());\n    };\n})();\n\n";
function euiHelper(skins) {
    return "\n    window.skins=window.skins||{};\n    window.generateEUI = {};\n    generateEUI.paths = {};\n    generateEUI.styles = undefined;\n    generateEUI.skins = " + JSON.stringify(skins) + ";\n    ";
}
