// @egret/eui-compiler/lib/emitter/json-emitter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var _1 = require(".");
var JSONEmitter = /** @class */ (function (_super) {
    __extends(JSONEmitter, _super);
    function JSONEmitter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.json = '';
        _this.euiNormalizeNames = {
            "$eBL": "eui.BitmapLabel",
            "$eB": "eui.Button",
            "$eCB": "eui.CheckBox",
            "$eC": "eui.Component",
            "$eDG": "eui.DataGroup",
            "$eET": "eui.EditableText",
            "$eG": "eui.Group",
            "$eHL": "eui.HorizontalLayout",
            "$eHSB": "eui.HScrollBar",
            "$eHS": "eui.HSlider",
            "$eI": "eui.Image",
            "$eL": "eui.Label",
            "$eLs": "eui.List",
            "$eP": "eui.Panel",
            "$ePB": "eui.ProgressBar",
            "$eRB": "eui.RadioButton",
            "$eRBG": "eui.RadioButtonGroup",
            "$eRa": "eui.Range",
            "$eR": "eui.Rect",
            "$eRAl": "eui.RowAlign",
            "$eS": "eui.Scroller",
            "$eT": "eui.TabBar",
            "$eTI": "eui.TextInput",
            "$eTL": "eui.TileLayout",
            "$eTB": "eui.ToggleButton",
            "$eTS": "eui.ToggleSwitch",
            "$eVL": "eui.VerticalLayout",
            "$eV": "eui.ViewStack",
            "$eVSB": "eui.VScrollBar",
            "$eVS": "eui.VSlider",
            "$eSk": "eui.Skin"
        };
        _this.elementContents = {};
        _this.elementIds = [];
        _this.skinParts = [];
        _this.states = {};
        _this.nodeMap = {};
        return _this;
    }
    JSONEmitter.prototype.getResult = function () {
        return this.json;
    };
    JSONEmitter.prototype.emitHeader = function (themeData) {
    };
    JSONEmitter.prototype.emitSkinNode = function (filename, skinNode) {
        var json = {};
        this.elementContents = {};
        this.elementIds = [];
        this.skinParts = [];
        this.states = [];
        this.nodeMap = {};
        var key = skinNode.namespace ? skinNode.namespace + "." + skinNode.classname : skinNode.classname;
        var item = {};
        json[key] = item;
        this.nodeMap[key] = skinNode;
        this.setPath(filename, item);
        this.setBaseState(skinNode, item);
        Object.assign(item, this.elementContents);
        if (this.skinParts.length > 0) {
            item['$sP'] = this.skinParts;
        }
        this.setStates(skinNode, item);
        item['$sC'] = "$eSk";
        this.json = JSON.stringify(json, null, 4);
    };
    JSONEmitter.prototype.setPath = function (filename, json) {
        json['$path'] = filename;
    };
    JSONEmitter.prototype.setBaseState = function (node, json, key) {
        if (key === void 0) { key = '$bs'; }
        var base = {};
        json[key] = base;
        for (var _i = 0, _a = node.attributes; _i < _a.length; _i++) {
            var attr = _a[_i];
            base[attr.key] = this.parseValue(attr.value);
        }
        if (node["type"]) {
            base['$t'] = this.convertType(node["type"]);
        }
        var elementContents = [];
        var sIds = [];
        for (var _b = 0, _c = node.children; _b < _c.length; _b++) {
            var child = _c[_b];
            var id = this.parseNode(child);
            this.hasAddType(child) ? sIds.push(id) : elementContents.push(id);
            this.setBaseState(child, this.elementContents, id);
        }
        elementContents.length > 0 && (base['$eleC'] = elementContents);
        sIds.length > 0 && (base['$sId'] = sIds);
    };
    JSONEmitter.prototype.setStates = function (skinNode, json) {
        if (skinNode.states.length === 0) {
            return;
        }
        json['$s'] = {};
        for (var _i = 0, _a = skinNode.states; _i < _a.length; _i++) {
            var state = _a[_i];
            json['$s'][state] = {};
        }
        this.getStatesAttribute(skinNode, json["$s"]);
    };
    JSONEmitter.prototype.getStatesAttribute = function (node, json) {
        var target = this.getNodeId(node);
        for (var _i = 0, _a = node.stateAttributes; _i < _a.length; _i++) {
            var attr = _a[_i];
            switch (attr.type) {
                case 'set':
                    {
                        if (attr.name in json) {
                            if (!json[attr.name]['$ssP']) {
                                json[attr.name]['$ssP'] = [];
                            }
                            json[attr.name]['$ssP'].push({
                                target: target,
                                name: attr.attribute.key,
                                value: attr.attribute.value
                            });
                        }
                    }
                    break;
                case 'add':
                    {
                        if (attr.name in json) {
                            if (!json[attr.name]['$saI']) {
                                json[attr.name]['$saI'] = [];
                            }
                            json[attr.name]['$saI'].push({
                                target: target,
                                property: "",
                                position: 1,
                                relativeTo: ""
                            });
                        }
                    }
                    break;
            }
        }
        for (var _b = 0, _c = node.children; _b < _c.length; _b++) {
            var child = _c[_b];
            this.getStatesAttribute(child, json);
        }
    };
    JSONEmitter.prototype.hasAddType = function (node) {
        for (var _i = 0, _a = node.stateAttributes; _i < _a.length; _i++) {
            var attr = _a[_i];
            if (attr.type === 'add') {
                return true;
            }
        }
        return false;
    };
    JSONEmitter.prototype.getNodeId = function (node) {
        var nodeMap = this.nodeMap;
        for (var id in nodeMap) {
            if (nodeMap[id] === node) {
                return id;
            }
        }
        return null;
    };
    JSONEmitter.prototype.parseNode = function (node) {
        var id = node.id;
        if (id) {
            this.skinParts.push(id);
        }
        else {
            var i = 1;
            var type = node.type.split('.').pop();
            do {
                id = "_" + type + i++;
            } while (this.elementIds.indexOf(id) !== -1);
        }
        this.elementIds.push(id);
        this.nodeMap[id] = node;
        return id;
    };
    JSONEmitter.prototype.parseValue = function (value) {
        if (!value["attributes"] && !value["children"]) {
            return value;
        }
        if (value["type"]) {
            var id = this.parseNode(value);
            this.setBaseState(value, this.elementContents, id);
            return id;
        }
        return value;
    };
    JSONEmitter.prototype.convertType = function (type) {
        for (var key in this.euiNormalizeNames) {
            if (this.euiNormalizeNames[key] === type) {
                return key;
            }
        }
        return "$eSk";
    };
    return JSONEmitter;
}(_1.BaseEmitter));
exports.JSONEmitter = JSONEmitter;
